crates/matrix-sdk/src/sliding_sync.rs

```diff
@@ -0,0 +1,957 @@
+    ) -> anyhow::Result<UpdateSummary> {
+        self.client.process_sliding_sync(resp.clone()).await?;
+        tracing::info!("main client processed.");
+        self.pos.replace(Some(resp.pos));
+        let mut updated_views = Vec::new();
+        if resp.lists.len() != views.len() {
+            bail!("Received response for {} lists, yet we have {}", resp.lists.len(), views.len());
+        }
+
+        for (view, updates) in views.iter().zip(resp.lists.iter()) {
+            let count: u32 =
+                updates.count.try_into().expect("the list total count convertible into u32");
+            tracing::trace!("view {:?}  update: {:?}", view.name, !updates.ops.is_empty());
+            if !updates.ops.is_empty() && view.handle_response(count, &updates.ops)? {
+                updated_views.push(view.name.clone());
+            }
+        }
+
+        let mut rooms = Vec::new();
+        let mut rooms_map = self.rooms.lock_mut();
+        for (id, room_data) in resp.rooms.iter() {
+            if let Some(mut r) = rooms_map.remove(id) {
+                r.update(room_data);
+                rooms_map.insert_cloned(id.clone(), r);
+                rooms.push(id.clone());
+            } else {
+                rooms_map.insert_cloned(
+                    id.clone(),
+                    SlidingSyncRoom::from(id.clone(), room_data.clone()),
+                );
+                rooms.push(id.clone());
+            }
+        }
+
+        Ok(UpdateSummary { views: updated_views, rooms })
+    }
+
+    /// Create the inner stream for the view.
+    ///
+    /// Run this stream to receive new updates from the server.
+    pub async fn stream<'a>(
+        &self,
+    ) -> anyhow::Result<impl Stream<Item = anyhow::Result<UpdateSummary>> + '_> {
+        let views = self.views.lock_ref().to_vec();
+        let _pos = self.pos.clone();
+
+        // FIXME: hack for while the sliding sync server is on a proxy
+        let client = self.client.clone();
+        let req_config =
+            self.homeserver.as_ref().map(|hs| RequestConfig::default().homeserver(hs.to_string()));
+
+        Ok(async_stream::try_stream! {
+            let mut remaining_views = views.clone();
+            let mut remaining_generators: Vec<SlidingSyncViewRequestGenerator<'_>> = views
+                .iter()
+                .map(SlidingSyncView::request_generator)
+                .collect();
+            loop {
+                let requests;
+                (requests, remaining_generators, remaining_views) = remaining_generators
+                    .into_iter()
+                    .zip(remaining_views)
+                    .fold(
+                    (Vec::new(), Vec::new(), Vec::new()), |mut c, (mut g, v)| {
+                        if let Some(r) = g.next() {
+                            c.0.push(r);
+                            c.1.push(g);
+                            c.2.push(v);
+                        }
+                        c
+                    });
```

Comment:
I'm usually all for combinators, but this seems like it would be much more readable with a few mutable `Vec`s declared first, and a regular `for` loop used to populate them.